projectq/setups/decompositions/cnu2toffoliandcu_test.py

```diff
@@ -0,0 +1,125 @@
+    qubit = eng.allocate_qubit()
+    ctrl_qureg = eng.allocate_qureg(2)
+    ctrl_qureg2 = eng.allocate_qureg(3)
+    eng.flush()
+    with Control(eng, ctrl_qureg):
+        Ph(0.1) | qubit
+        Ry(0.2) | qubit
+    with Control(eng, ctrl_qureg2):
+        QFT | qubit + ctrl_qureg
+        X | qubit
+    eng.flush()  # To make sure gates arrive before deallocate gates
+    eng.flush(deallocate_qubits=True)
+    # Don't test initial 6 allocate and flush and trailing deallocate
+    # and two flush gates.
+    for cmd in saving_backend.received_commands[7:-8]:
+        assert cnu2toffoliandcu._recognize_CnU(cmd)
+
+
+def test_recognize_incorrect_gates():
+    saving_backend = DummyEngine(save_commands=True)
+    eng = MainEngine(backend=saving_backend)
+    qubit = eng.allocate_qubit()
+    ctrl_qubit = eng.allocate_qubit()
+    ctrl_qureg = eng.allocate_qureg(2)
+    eng.flush()
+    with Control(eng, ctrl_qubit):
+        Rx(0.3) | qubit
+    X | qubit
+    with Control(eng, ctrl_qureg):
+        X | qubit
+    eng.flush(deallocate_qubits=True)
+    for cmd in saving_backend.received_commands:
+        assert not cnu2toffoliandcu._recognize_CnU(cmd)
+
+
+def _decomp_gates(eng, cmd):
+    g = cmd.gate
+    if isinstance(g, ClassicalInstructionGate):
+        return True
+    if len(cmd.control_qubits) <= 1:
+        return True
+    if len(cmd.control_qubits) == 2 and isinstance(cmd.gate, XGate):
+        return True
+    return False
+
+
+def test_decomposition():
+    for basis_state_index in range(0,16):
+        basis_state = [0] * 16
+        basis_state[basis_state_index] = 1.
+        correct_eng = MainEngine(backend=Simulator(), engine_list=[])
+
+        rule_set = DecompositionRuleSet(modules=[cnu2toffoliandcu])
+        test_eng = MainEngine(backend=Simulator(),
+                              engine_list=[AutoReplacer(rule_set),
+                                           InstructionFilter(_decomp_gates)])
+        test_sim = test_eng.backend
+        correct_sim = correct_eng.backend
+        correct_qb = correct_eng.allocate_qubit()
+        correct_ctrl_qureg = correct_eng.allocate_qureg(3)
+        correct_eng.flush()
+        test_qb = test_eng.allocate_qubit()
+        test_ctrl_qureg = test_eng.allocate_qureg(3)
+        test_eng.flush()
+
+        correct_sim.set_wavefunction(basis_state, correct_qb +
+                                     correct_ctrl_qureg)
+        test_sim.set_wavefunction(basis_state, test_qb + test_ctrl_qureg)
+
+        with Control(test_eng, test_ctrl_qureg[:2]):
+            Rx(0.4) | test_qb
+        with Control(test_eng, test_ctrl_qureg):
+            Ry(0.6) | test_qb
+
+        with Control(correct_eng, correct_ctrl_qureg[:2]):
+            Rx(0.4) | correct_qb
+        with Control(correct_eng, correct_ctrl_qureg):
+            Ry(0.6) | correct_qb
+
+        test_eng.flush()
+        correct_eng.flush()
+
+        for fstate in ['0000', '0001', '0010', '0011', '0100', '0101', '0110',
```

Comment:
Why not `for fstate in range(16)` and then pass `bin(fstate)[2:]` to get_amplitude?

Reply:
👍  Done.